### PR TITLE
file: fflush all output streams on fflush(NULL)

### PR DESCRIPTION
**Pull request by agkaminski**
DONE: RTOS-494

## Description

Fixes https://github.com/phoenix-rtos/phoenix-rtos-project/issues/734

## Types of changes

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## How Has This Been Tested?

- [ ] Already covered by automatic testing.
- [ ] New test added: (add PR link here).
- [ ] Tested by hand on: (list targets here).

## Checklist:

- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have added tests to cover my changes.
- [x] All new and existing linter checks and tests passed.
- [x] My changes generate no new compilation warnings for any of the targets.

## Special treatment

- [ ] This PR needs additional PRs to work (list the PRs, preferably in merge-order).
- [ ] I will merge this PR by myself when appropriate.

### CODE REVIEW

File: stdio/file.c
Comment (nalajcie) [inline stdio/file.c:626]: maybe move to separate func `fflush_single`  or move stream == NULL semantics to `fflush` for better readability (and avoid recursion-at-a-glance)?

File: stdio/file.c
Comment (agkaminski) [reply stdio/file.c:615]: Yeah, but it's stated that for `fflush(NULL)` _only_ **output** streams should be `fflush`ed. Or do am I missing something?

File: stdio/file.c
Comment (agkaminski) [reply stdio/file.c:626]: I'd rather not create a new function, really, less better.

Hmmm, I placed this in unlocked to handle `fflush_unlocked(NULL`) too, but I guess I handle it wrong, as I get locks anyway. I wonder, should `fflush_unlocked(NULL)` work too?

Comment (agkaminski) [reply stdio/file.c:626]: I don't see it stated explicitly, but I guess `fflush_unlocked(NULL)` should be handled

Comment (agkaminski) [reply stdio/file.c]: "I'd rather not", but in the end I added not one but two functions, lol

Comment (nalajcie) [reply stdio/file.c]: hehh, `fflush_unlocked` is linux-specific and implemented only in `glibc` so IMHO we don't need to focus on correctness here. I've checked the `glibc` implementation and `fflush_unlocked(NULL)` is the same as `fflush(NULL)` (acquires per-file locks - https://github.com/lattera/glibc/blob/master/libio/iofflush_u.c ). Surprisingly - there is an unlocked version which is used in cleanup (`fcloseall` implemetation and probably destructors).

I think You can either leave it as-is or remove unlocked NULL version for code simplicity.

File: stdio/file.c
Comment (nalajcie) [inline stdio/file.c:152]: first remove from the list, then start destroying. Right now there might be a race with `fflush(NULL)` which will try to acquire destroyed lock?

**Review by nalajcie — APPROVED** [review]

File: stdio/file.c
Comment (lukileczo) [inline stdio/file.c:609]: Why not merge this into one `if`?

Comment (agkaminski) [reply stdio/file.c:609]: I don't want to go to `else` if `stream->bufpos == 0`

Comment (agkaminski) [reply stdio/file.c:609]: I don't want to go to else if stream->bufpos == 0

Comment (lukileczo) [reply stdio/file.c:609]: okay, ofc

**Review by lukileczo — APPROVED** [review]

**Review by adamdebek — APPROVED** [review]
Now test case with flushed streams in exit() test passes.

**Pull request merged**
